src/kbs2.jl

```diff
@@ -13,17 +13,18 @@ function deriverule!(rs::RewritingSystem, stack, o::Ordering = ordering(rs))
             if lt(o, a, b)
                 a, b = b, a
             end
-            push!(rs, a => b)
 
-            for i in 1:length(rules(rs))-1
-                (lhs, rhs) = rules(rs)[i]
-                if isactive(rs, i) && occursin(a, lhs)
+            for (i, (lhs, rhs)) in enumerate(rules(rs))
+                isactive(rs, i) || continue
+                if occursin(a, lhs)
                     setinactive!(rs, i)
                     push!(stack, lhs => rhs)
                 else occursin(a, rhs)
                     rules(rs)[i] = (lhs => rewrite_from_left(rhs, rs))
                 end
             end
+
+            push!(rs, a => b)
```

Comment:
Again I think this is wrong. In the former one we first did `push!(rs, a=> b)` and then inside the for-loop we did rewriting wrt. to the RewritingSystem with that additional rule. A priori this yield different results.

Reply:
it now seems to me that `else` statement is incorrect; what we should do is to first rewrite `rhs → rhs_new` with rule `a => b` (we know that this is the only rewriting rule that applies at the moment), then we need to add `a=>b` to `rs` and finally call `rewrite_from_left(rhs, rs)`.

Reply:
I think it will still be incorrect: noticing that in the beginning `push!(rs, a => b)` was outside the loop I believe your version may happen to push the the rule `a => b` multiple times to the RewritingSystem (because it may happen that `else` will be triggered multiple times in the for-loop).

Reply:
I thought rather of something like this:
```julia
function rewrite_from_left!(
    v::AbstractWord,
    w::AbstractWord,
    rule::Pair{<:AbstractWord, <:AbstractWord},
)
    lhs, rhs = rule
    while !isone(w)
        push!(v, popfirst!(w))
        if issuffix(lhs, v)
            prepend!(w, rhs)
            resize!(v, length(v) - length(lhs))
        end
    end
    return v
end

function rewrite_from_left(u::W, rule::Pair{<:AbstractWord, <:AbstractWord}) where {W<:AbstractWord}
    T = eltype(u)
    v = BufferWord{T}(0, length(u))
    w = BufferWord{T}(u, 0, 0)
    v = rewrite_from_left!(v, w, rule)
    return W(v)
end
```

then the `else` block (btw. there should be `elseif`, instead of `else`!)
```julia
                elseif occursin(a, rhs)
                    rule = a => b
                    new_rhs = rewrite_from_left(rhs, rule)
                    last(rules(rs)) != rule && push!(rs, rule)
                    rules(rs)[i] = (lhs => rewrite_from_left(new_rhs, rs))
                end
```

Reply:
Ahhh. Then its OK :D